File: player.py
import discord
import asyncio
from config import DEFAULT_VOLUME


class MusicPlayer:
    def __init__(self, bot, guild):
        self.bot = bot
        self.guild = guild
        self.queue = []
        self.history = []
        self.volume = DEFAULT_VOLUME
        self.loop_mode = "off"
        self.voice = None
        self.current = None
        self.start_time = None
        self.twenty_four_seven = False

    async def connect(self, channel):
        if not self.voice or not self.voice.is_connected():
            self.voice = await channel.connect()

    async def add_song(self, song, ctx=None):
        self.queue.append(song)

        if self.current and ctx:
            await ctx.send(f"✅ **Next Song Added:** {song['title']}")

    async def play_next(self):
        if not self.voice:
            return

        if self.current:
            self.history.append(self.current)

        if self.loop_mode == "song" and self.current:
            pass
        else:
            if self.loop_mode == "queue" and self.current:
                self.queue.append(self.current)

            if self.queue:
                self.current = self.queue.pop(0)
            else:
                if not self.twenty_four_seven:
                    await self.voice.disconnect()
                    self.voice = None
                return

        # ✅ SPEED LOCKED AT 1x
        ffmpeg_options = {
            "before_options": "-reconnect 1 -reconnect_streamed 1 -reconnect_delay_max 5",
            "options": "-vn -filter:a atempo=1.0"
        }

        source = discord.FFmpegPCMAudio(
            self.current["url"],
            **ffmpeg_options
        )

        source = discord.PCMVolumeTransformer(
            source,
            volume=self.volume
        )

        self.start_time = asyncio.get_event_loop().time()

        self.voice.play(
            source,
            after=lambda e: asyncio.run_coroutine_threadsafe(
                self.play_next(),
                self.bot.loop
            )
        )

    def get_progress(self):
        if not self.start_time:
            return 0
        return int(asyncio.get_event_loop().time() - self.start_time)

File: player.js
const { 
    joinVoiceChannel, 
    createAudioPlayer, 
    createAudioResource, 
    AudioPlayerStatus 
} = require('@discordjs/voice');

const play = require('play-dl');
const { DEFAULT_VOLUME } = require('./config');

class MusicPlayer {
    constructor(client, guild) {
        this.client = client;
        this.guild = guild;
        this.queue = [];
        this.history = [];
        this.volume = DEFAULT_VOLUME || 0.5;
        this.loopMode = "off";
        this.connection = null;
        this.player = createAudioPlayer();
        this.current = null;
        this.startTime = null;
        this.twentyFourSeven = false;

        // ✅ Auto play next
        this.player.on(AudioPlayerStatus.Idle, () => {
            this.playNext().catch(console.error);
        });

        // ✅ Prevent crash
        this.player.on("error", (error) => {
            console.error("Audio Player Error:", error);
            this.playNext().catch(console.error);
        });
    }

    async connect(channel) {
        if (!channel) return;

        if (!this.connection) {
            this.connection = joinVoiceChannel({
                channelId: channel.id,
                guildId: channel.guild.id,
                adapterCreator: channel.guild.voiceAdapterCreator,
                selfDeaf: true
            });

            this.connection.subscribe(this.player);
        }
    }

    async addSong(song) {
        if (!song || !song.url) {
            console.log("⚠ Invalid song skipped (no URL)");
            return;
        }

        this.queue.push(song);
    }

    async playNext() {
        try {

            if (this.current)
                this.history.push(this.current);

            if (this.loopMode === "song" && this.current) {
                // replay same
            } else {

                if (this.loopMode === "queue" && this.current) {
                    this.queue.push(this.current);
                }

                if (this.queue.length > 0) {
                    this.current = this.queue.shift();
                } else {
                    this.current = null;

                    if (!this.twentyFourSeven && this.connection) {
                        this.connection.destroy();
                        this.connection = null;
                    }
                    return;
                }
            }

            // ✅ SAFETY CHECK (IMPORTANT)
            if (!this.current || !this.current.url) {
                console.log("⚠ Skipping invalid song (no URL)");
                return this.playNext();
            }

            const stream = await play.stream(this.current.url);

            const resource = createAudioResource(stream.stream, {
                inputType: stream.type,
                inlineVolume: true
            });

            resource.volume.setVolume(this.volume);

            this.startTime = Date.now();

            this.player.play(resource);

        } catch (error) {
            console.error("PlayNext Error:", error);
            return this.playNext();
        }
    }

    getProgress() {
        if (!this.startTime) return 0;
        return Math.floor((Date.now() - this.startTime) / 1000);
    }
}

module.exports = MusicPlayer;
